chain/client/src/sync.rs

```diff
@@ -685,24 +658,29 @@ impl StateSync {
                     shard_id,
```

Comment:
I suggest we do not make an implicit `Option<T>` here, but instead make it explicit for readability and safety from bugs reason. We can either wrap `part_ids` with `num_parts` into a new structure `Parts` or we can create an alias `type Parts = (Vec<u64>, u64)`, then have a `Option<Parts>` field in `StateRequest`, also do the same for `ShardStateSyncResponse`. This will helps whoever reads the code understand that this structure has two use cases: to download headers only and to get parts.

Reply:
Make sense to do it while refactoring later.
I also afraid that `type` cannot be serialized and deserialized which may be painful. See commit `9322869c7` for example.

Reply:
> I also afraid that `type` cannot be serialized and deserialized which may be painful. 

Serialization/deserialization is oblivious of `type` alias, it is pure syntactic sugar.

> See commit 9322869c7 for example.

Which line should I look at in https://github.com/nearprotocol/nearcore/commit/9322869c7 ?

> Make sense to do it while refactoring later.

I would have agreed if this code was not threading changes through our entire code base. We should keep the number of PRs that thread through the entire codebase small. I already had issues with parts implementation breaking genesis tools.

Reply:
> Which line should I look at in 9322869 ?
<img width="540" alt="Screenshot 2019-11-16 at 10 51 38" src="https://user-images.githubusercontent.com/5685493/68986899-19716a80-085f-11ea-9dd1-b3e58defddcb.png">
The reason of changing `PartialState` from `type` is because it cannot be serialized with borsh. (Or maybe I have not enough knowledge in rust how to `derive` for types).

Reply:
> I already had issues with parts implementation breaking genesis tools.

Is it related to not having num_parts in genesis file? This PR will fix that issue.

Reply:
> I would have agreed if this code was not threading changes through our entire code base. We should keep the number of PRs that thread through the entire codebase small.

It sounds reasonable. However, I still don't see the advantages of having more detailed structure.

> This will helps whoever reads the code understand that this structure has two use cases: to download headers only and to get parts.

It may be subjective.

Reply:
> The reason of changing `PartialState` from `type` is because it cannot be serialized with borsh. (Or maybe I have not enough knowledge in rust how to `derive` for types).

There should be no need in doing that. If `StateItem` implements borsh then 
`pub type PartialState = Vec<StateItem>` implements it too.

> It sounds reasonable. However, I still don't see the advantages of having more detailed structure.
> It may be subjective.

The point of using strong explicit types instead of encoding `Option` implicitly is that strong typing reducing the number of potential bugs and makes readability simpler. I don't think anyone argues that using `enum` is better than encoding variants with integers C-style, because: it adds readability and prevents people from incorrectly using integers. 

When someone sees somewhere in the code `(Vec<u64>, u64)` they need to trace  all the way through the code to the origin of this structure in order to understand what this type represents. This is time-consuming.

Reply:
> There should be no need in doing that. 

That should not be true. I got compilation error because of serialization isn't derived for types.

It still may be my fault.

Reply:
> When someone sees somewhere in the code (Vec<u64>, u64) they need to trace all the way through the code to the origin of this structure in order to understand what this type represents. This is time-consuming.

It hardly makes sense. However, I don't want to argue here anymore. Sometimes arguing is time-consuming.

++